### PR TITLE
[ADD] openacademy: Wizards. T#57454

### PR DESCRIPTION
**Pull request by ejdecena**
Exercise 29: Launch the wizard.
Exercise 28: Define the wizard.

### CODE REVIEW

File: wizards/openacademy_wizard.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from odoo import models, fields, api
+
+class Wizard(models.TransientModel):
+    _name = "openacademy.wizard"
+    _description = "This is the description of Wizard model."
```
Comment (luisg123v) [inline wizards/openacademy_wizard.py]:
```suggestion
    _description = "Openacademy Wizard"
```

File: models/res_partner.py [hunk 2]
```diff
@@ -11,3 +11,10 @@ class Partner(models.Model):
         column2="session_id",
         string="Sessions",
     )
+    wizard_ids = fields.Many2many(
```
Comment (luisg123v) [inline models/res_partner.py]: Remove, a field to a transient model makes no sense, it will never be used.

File: views/openacademy_wizard_views.xml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (luisg123v) [inline views/openacademy_wizard_views.xml]: - Missing this file on the manifest. As I mentioned on last meeting, it's mandatory to test everything before requesting review
- All wizard files (including views) should be placed under the "wizards" directory, as you did for the model

File: wizards/openacademy_wizard.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    session_id = fields.Many2one("openacademy.session", string="Session")
```
Comment (luisg123v) [inline wizards/openacademy_wizard.py]: - Remove superfluous string
- Add `required=True`, as without this the wizard couldn't work

**Review by luisg123v — CHANGES_REQUESTED** [review]

File: wizards/openacademy_wizard.py (hunk 4 above)
Comment (ejdecena) [reply wizards/openacademy_wizard.py]: All items done.

File: wizards/openacademy_wizard.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from odoo import models, fields, api
```
Comment (ajqn9094) [inline wizards/openacademy_wizard.py]: It is good not to make imports that we do not use in the code, this to avoid Lint problems.

I know that later you can end up using it, but since this is not the case at the moment it could be removed.

```suggestion
from odoo import models, fields
```

File: __init__.py [hunk 6]
```diff
@@ -1 +1,2 @@
 from . import models
+from . import wizards
```
Comment (ajqn9094) [inline __init__.py]: To maintain the Odoo standard I think the directory should be called "wizard".

```suggestion
from . import wizard
```

File: __manifest__.py [hunk 7]
```diff
@@ -13,6 +13,7 @@
         "views/openacademy_course_views.xml",
         "views/openacademy_session_views.xml",
         "views/res_partner_views.xml",
+        "wizards/openacademy_wizard_views.xml",
```
Comment (ajqn9094) [inline __manifest__.py]: Same case that above.

```suggestion
        "wizard/openacademy_wizard_views.xml",
```

File: __init__.py (hunk 6 above)
Comment (luisg123v) [reply __init__.py]: Quite the opposite. Plural, please, to conform OCA guidelines

Comment (ajqn9094) [reply __init__.py]: Sorry for re confusion.

File: wizards/openacademy_wizard.py (hunk 5 above)
Comment (ejdecena) [reply wizards/openacademy_wizard.py]: Got it.

File: __init__.py (hunk 6 above)
Comment (ejdecena) [reply __init__.py]: I am following the file naming of the OCA -> https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#file-naming
temp
.

File: wizards/openacademy_wizard.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    session_id = fields.Many2one("openacademy.session", required=True)
+
+    partner_ids = fields.Many2many(
+        comodel_name="res.partner",
+        relation="openacademy_wizard_partner_rel",
+        column1="wizard_id",
+        column2="partner_id",
+        string="Partners",
+    )
+
```
Comment (luisg123v) [inline wizards/openacademy_wizard.py]: Remove extra blank line at end of file.

I'd recommend you run `flake8` before pushing to avoid this. For instance, I just ran it over this branch and it output the following:

```
./models/openacademy_session.py:4:1: E302 expected 2 blank lines, found 1
./models/openacademy_session.py:38:47: E226 missing whitespace around arithmetic operator
./models/openacademy_session.py:38:68: E226 missing whitespace around arithmetic operator
./models/res_partner.py:3:1: E302 expected 2 blank lines, found 1
./models/openacademy_course.py:3:1: E302 expected 2 blank lines, found 1
./wizards/openacademy_wizard.py:3:1: E302 expected 2 blank lines, found 1
./wizards/openacademy_wizard.py:16:1: W391 blank line at end of file
```

Could you fix the wizard-related ones here and create a new PR to fix lints from other code, please?

File: wizards/openacademy_wizard_views.xml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        <field name="model">openacademy.wizard</field>
+        <field name="arch" type="xml">
+            <form string="Wizard">
+                <sheet>
+                    <group name="wizards" string="Wizard">
+                        <field name="session_id" />
+                        <field name="partner_ids" />
+                    </group>
+                </sheet>
+                <footer>
```
Comment (luisg123v) [inline wizards/openacademy_wizard_views.xml]: Missing button to actually perform an action

**Review by luisg123v — CHANGES_REQUESTED** [review]

File: views/openacademy_menu.xml [hunk 10]
```diff
@@ -25,7 +33,17 @@
 
     <menuitem id="menu_session"
         name="Sessions"
-        parent="menu_root"
+        parent="menu_root"/>
+
+    <menuitem id="sub_menu_session"
+        name="Sessions"
```
Comment (luisg123v) [inline views/openacademy_menu.xml]: This is the same name as above

File: wizards/openacademy_wizard.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    @api.model
+    def default_get(self, fields_list):
+        result = super().default_get(fields_list)
+        model = self.env.context.get("active_model")
+        if model == "openacademy.session":
+            session_id = self.env.context.get("active_id")
+            if session_id:
```
Comment (luisg123v) [inline wizards/openacademy_wizard.py]: Don't need two conditionals, just extract active model and ID before conditional and add two conditions

File: wizards/openacademy_wizard_views.xml [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+    <record id="wizard_view_form" model="ir.ui.view">
+        <field name="name">openacademy.wizard.view.form</field>
+        <field name="model">openacademy.wizard</field>
+        <field name="arch" type="xml">
+            <form string="Wizard">
+                <sheet>
+                    <group name="wizards" string="Wizard">
+                        <field name="session_id" />
+                        <field name="partner_ids" />
```
Comment (luisg123v) [inline wizards/openacademy_wizard_views.xml]: pro tip: you can use the widget `many2many_tags` to select them inline:

```suggestion
                        <field name="partner_ids" widget="many2many_tags" />
```

**Review by luisg123v — CHANGES_REQUESTED** [review]

File: models/openacademy_course.py [hunk 13]
```diff
@@ -10,8 +11,13 @@ class Course(models.Model):
     session_ids = fields.One2many("openacademy.session", string="Sessions", inverse_name="course_id")
 
     _sql_constraints = [
-        ("description_and_title_different", "CHECK(name != description)", "The title and description of the course must be different."),
-        ("unique_course_name", "UNIQUE(name)", "The name of the course must be unique.")
+        ("description_and_title_different",
```
Comment (luisg123v) [inline models/openacademy_course.py]: This is an unrelated change. Please open a new PR to address only lints fixing (as I commented).

**Review by luisg123v — CHANGES_REQUESTED** [review]

File: views/openacademy_menu.xml (hunk 10 above)
Comment (luisg123v) [reply views/openacademy_menu.xml]: Sorry, I just noticed one menu was inside the other, so both having the same name is expected.

File: wizards/openacademy_wizard.py (hunk 11 above)
Comment (luisg123v) [reply wizards/openacademy_wizard.py]: This was not applied as asked. Now there's only one conditional with one condition and active ID is always added

Comment (ejdecena) [reply wizards/openacademy_wizard.py]: It is applied, in the code is:
```python
@api.model
 def default_get(self, fields_list):
        result = super().default_get(fields_list)
        model = self.env.context.get("active_model")
        if model == "openacademy.session":
            session_id = self.env.context.get("active_id")
            result.update({"session_id": session_id})
        return result
```
This update here is not reflected in this PR.

Comment (luisg123v) [reply wizards/openacademy_wizard.py]: In that code, you're updating session_id even if active_id is False

File: openacademy/models/openacademy_course.py [hunk 14]
```diff
@@ -10,8 +11,13 @@ class Course(models.Model):
     session_ids = fields.One2many("openacademy.session", string="Sessions", inverse_name="course_id")
 
     _sql_constraints = [
-        ("description_and_title_different", "CHECK(name != description)", "The title and description of the course must be different."),
-        ("unique_course_name", "UNIQUE(name)", "The name of the course must be unique.")
+        ("description_and_title_different",
```
Comment (luisg123v) [inline openacademy/models/openacademy_course.py]: This is an unrelated change, this will be done in #11

File: wizards/openacademy_wizard.py (hunk 11 above)
Comment (luisg123v) [reply wizards/openacademy_wizard.py]: This is not applied yet. Could you check the example code I provided, please?

As I mentioned, here you are updating default value even if active IDs are not present.

You should extract active ID(s) and active model. If active ID is not set or active model doesn't match, you do nothing.

File: wizards/openacademy_wizard_views.xml (hunk 12 above)
Comment (luisg123v) [reply wizards/openacademy_wizard_views.xml]: This change was reverted

File: openacademy/wizards/openacademy_wizard_views.xml [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+    <record id="wizard_view_form" model="ir.ui.view">
+        <field name="name">openacademy.wizard.view.form</field>
+        <field name="model">openacademy.wizard</field>
+        <field name="arch" type="xml">
+            <form string="Wizard">
+                <sheet>
+                    <group name="wizards" string="Wizard">
+                        <field name="session_ids" />
+                        <field name="partner_ids"/>
```
Comment (luisg123v) [inline openacademy/wizards/openacademy_wizard_views.xml]: Use also here the widget `many2many_tags`

**Review by luisg123v — CHANGES_REQUESTED** [review]

File: openacademy/wizards/openacademy_wizard_views.xml (hunk 15 above)
Comment (ejdecena) [reply openacademy/wizards/openacademy_wizard_views.xml]: But, how could **multiple** attendees be added if not through a tree view?
In the Vauxoo tutorial a tree view is used for add the attendees -> https://vauxoo.thinkific.com/courses/take/odoo-backend/lessons/1966754-ejercicio-asistentes

Comment (ejdecena) [reply openacademy/wizards/openacademy_wizard_views.xml]: I really don't understand you, sorry. The following code is the one that is present in this PR, and it works:
```python
    @api.model
    def default_get(self, fields_list):
        result = super().default_get(fields_list)
        model = self.env.context.get("active_model")
        if model == "openacademy.session":
            session_ids = self.env.context.get("active_ids")
            sessions = self.env.get(model).browse(session_ids)
            result.update({"session_ids": sessions})
        return result
```

Comment (luisg123v) [reply openacademy/wizards/openacademy_wizard_views.xml]: I mean:

```
@api.model
def default_get(self, fields_list):
    result = super().default_get(fields_list)
    active_model = self.env.context.get("active_model")
    active_ids = self.env.context.get("active_ids")
    if active_model == "openacademy.session" and active_ids:
        result["session_ids"] = [(6, 0, active_ids)]
    return result
```

Comment (ejdecena) [reply openacademy/wizards/openacademy_wizard_views.xml]: Okay, I ask something, when the `active_model = "openacademy.session"`, doesn't this determine that the `active_ids` is already set? In the worst case the `active_ids` would be an empty list.

The notation `[(6, 0, active_ids)]` doesn't look very clear to me, we haven't seen it in the tutorial.

Comment (ejdecena) [reply openacademy/wizards/openacademy_wizard_views.xml]: A fix or improvement could be this:
```python
    @api.model
    def default_get(self, fields_list):
        result = super().default_get(fields_list)
        model = self.env.context.get("active_model")
        session_ids = self.env.context.get("active_ids")
        if model == "openacademy.session" and session_ids:
            sessions = self.env.get(model).browse(session_ids)
            result.update({"session_ids": sessions})
        return result
```

Comment (luisg123v) [reply openacademy/wizards/openacademy_wizard_views.xml]: > In the worst case the active_ids would be an empty list.

In this case, yes. But it's a good practice to write code as robust as possible. In the future, you could implement wizards that are called not only from the user interface, but also by code.

> The notation [(6, 0, active_ids)] doesn't look very clear to me

In deed, it isn't. That's why in v15 it was replaced by [the Command class](https://www.odoo.com/documentation/15.0/developer/reference/backend/orm.html#odoo.fields.Command). Unfortunately, the old syntax is still all around the code (people don't even know it yet), so you will need to get familiarized with ithe old syntax, you can read [the doc of v14](https://www.odoo.com/documentation/14.0/developer/reference/addons/orm.html##create-update). For this specific case, it would be:

> (6, 0, ids)
> 
>     replaces all existing records in the set by the ids list

> A fix or improvement could be this:

In your code, you're browsing Ids, which creates a recordset, but only IDs are used in the end. Besides, it's generally a better idea to return Ids when dealing with dictionaries, as they work in both code and views (managed by frontend). Maybe not as necessary on newer versions, though.

Comment (ejdecena) [reply openacademy/wizards/openacademy_wizard_views.xml]: Okay, I understand. Thank you.

File: models/openacademy_course.py (hunk 13 above)
Comment (ejdecena) [reply models/openacademy_course.py]: This change comes from the rebasing with the 15.0 branch.

File: views/openacademy_menu.xml (hunk 10 above)
Comment (luisg123v) [reply views/openacademy_menu.xml]: This is still missing

**Review by luisg123v — APPROVED** [review]
